### DISCUSSION TITLE
hist2dplot on matplotlib axis breaks

### DISCUSSION THREAD

**Discussion by MoAly98** [scikit-hep/mplhep#368] [Q&A]
Hi, 

I'm trying to plot a 2D-Histogram using `hist2dplot` on an existing matplotlib axis. Following the example on the README
```
xbins, ybins = [0, 1, 2, 3], [0, 1, 2, 3]
H = np.array([[2,3,2], [1,2,1], [3,1,3]])
hep.hist2dplot(H, xbins, ybins)
```
works, but if i have a matplotlib axis `axis` and I execute:
```
xbins, ybins = [0, 1, 2, 3], [0, 1, 2, 3]
H = np.array([[2,3,2], [1,2,1], [3,1,3]])
hep.hist2dplot(H, xbins, ybins, ax= axis)
```
this breaks giving the following error

```
  File "/afs/cern.ch/user/m/maly/.conda/envs/xbb_env/lib/python3.6/site-packages/mplhep/plot.py", line 505, in hist2dplot
    plt.sca(ax)
  File "/afs/cern.ch/user/m/maly/.conda/envs/xbb_env/lib/python3.6/site-packages/matplotlib/pyplot.py", line 984, in sca
    _pylab_helpers.Gcf.set_active(ax.figure.canvas.manager)
  File "/afs/cern.ch/user/m/maly/.conda/envs/xbb_env/lib/python3.6/site-packages/matplotlib/_pylab_helpers.py", line 127, in set_active
    cls.figs[manager.num] = manager
AttributeError: 'NoneType' object has no attribute 'num'
```
 I checked the type of `axis` to make sure it is not None, and indeed it is`AxesSubplot(0.125,0.208882;0.562413x0.48434)`

Am I doing something wrong here? 

Thanks,
Mo

**Comment by MoAly98** [comment]
This is with `mplhep` version `0.2.18`

**Reply by andrzejnovak** [reply]
Can you show the rest of the code? Seems to work for a simple ax

![image](https://user-images.githubusercontent.com/13226500/157448582-06c23fef-62d5-4b5e-9164-85f1937daed2.png)

**Reply by MoAly98** [reply]
I realised that version `0.2.18`is too old. I upgraded to `0.3.20` and I get a different error:
```
  File "/opt/anaconda3/envs/pythium/lib/python3.9/site-packages/mplhep/plot.py", line 529, in hist2dplot
    plt.sca(ax)
  File "/opt/anaconda3/envs/pythium/lib/python3.9/site-packages/matplotlib/pyplot.py", line 1085, in sca
    figure(ax.figure)
  File "/opt/anaconda3/envs/pythium/lib/python3.9/site-packages/matplotlib/pyplot.py", line 753, in figure
    raise ValueError("The passed figure is not managed by pyplot")
ValueError: The passed figure is not managed by pyplot
```

when I run the following (more elaborate) code:
```
from matplotlib.figure import Figure
from matplotlib import gridspec
from matplotlib.backends.backend_agg import FigureCanvasAgg as FigureCanvas
import mplhep as hep

fig = Figure((15,15))
can = FigureCanvas(fig)
gs = gridspec.GridSpec(5, 3, height_ratios=(0.7, 1,1 ,0.1,0.1), width_ratios = (1,1,0.7))
ax = fig.add_subplot(gs[1:3, :2])
H = np.array([[2,3,2], [1,2,1], [3,1,3]])
xbins, ybins = [0, 1, 2, 3], [0, 1, 2, 3]
hep.hist2dplot(H, xbins, ybins, ax= ax)
```

I produce the axis this way because of some class structure that I have. It is also worth noting that if i replace the last line with
```
 hep.histplot(H[:,0], xbins, ax= ax)
```

the code  runs without any errors. The problem seems to be just for `hist2dplot`.

**Reply by andrzejnovak** [reply]
Ok, I think this is actually `matplotlib` question. The error you are getting doesn't seem to interact with the hist2dplot method really. 

I am not sure of the canvas purpose here, but removing that line and initiating the figure with `fig = plt.figure(figsize=(15,15))` seems to behave normally for me
